### PR TITLE
Refactors bitflag macro values into bitshifts

### PR DESCRIPTION
**Pull request by ohnopigeons**
I'd like this format to be the standard but before that happens existing code needs to be standardized first.

This bitshift format:
Prevents miscalculation/typo errors (that have happened in the past)
Makes it easier to account for bitflags and their positions
Makes it easier to ensure no bitflag exceeds byond's 16-bit bitfield limit

### CODE REVIEW

**Comment by pigeonsk** [comment]
There's some formatting issues with my editor I'll try to get that fixed.

File: code/__DEFINES/components.dm [hunk 1]
```diff
@@ -6,10 +6,10 @@
...
 // How multiple components of the exact same type are handled in the same datum
 
-#define COMPONENT_DUPE_HIGHLANDER 0		//old component is deleted (default)
-#define COMPONENT_DUPE_ALLOWED 1		//duplicates allowed
-#define COMPONENT_DUPE_UNIQUE 2			//new component is deleted
-#define COMPONENT_DUPE_UNIQUE_PASSARGS 4	//old component is given the initialization args of the new
+#define COMPONENT_DUPE_HIGHLANDER 		0		//old component is deleted (default)
+#define COMPONENT_DUPE_ALLOWED			(1<<0)	//duplicates allowed
+#define COMPONENT_DUPE_UNIQUE 			(1<<1)	//new component is deleted
+#define COMPONENT_DUPE_UNIQUE_PASSARGS 	(1<<2)	//old component is given the initialization args of the new
```
Comment (SpaceManiac) [inline code/__DEFINES/components.dm]: These aren't bitflags, it's just a sequence which happens to skip 3.

Comment (pigeonsk) [reply code/__DEFINES/components.dm]: Why is it skipping 3?

**Comment by arsserpentarium** [comment]
Errr...So, for every bitflag, byond will perform bitshift,every time it's used?

**Comment by pigeonsk** [comment]
It should be evaluated as a constant at compile time but admittedly I'm not an expert in byond or its compiler.

**Comment by Cyberboss** [comment]
It's more the cost of running the op through the VM


On Apr 15, 2018 5:21 PM, "pigeons" <notifications@github.com> wrote:

You are talking about literal microseconds here, if not nanoseconds.

—
You are receiving this because your review was requested.
Reply to this email directly, view it on GitHub
<https://github.com/tgstation/tgstation/pull/37234#issuecomment-381439021>,
or mute the thread

**Comment by pigeonsk** [comment]
Bit shifts are such a fundamental machine instruction I'd be surprised if it was twice as expensive or more on a VM.

**Review by duncathan — APPROVED** [review]

**Comment by pubby** [comment]
byond does have constant folding

http://www.byond.com/forum/?post=1305354

**Comment by vuonojenmustaturska** [comment]
but does it work for bit shifts? bit shifts in byond are not a pure binary operation, but rather a multi-step operation involving a value check, conversion to int, the operation itself and a conversion to float

**Comment by pigeonsk** [comment]
Where are you getting the conversions from?

**Comment by ExcessiveUseOfCobblestone** [comment]
you could always just do (2^0) (2^1) to get the same effect lol.

**Comment by optimumtact** [comment]
this is a bikeshed formatting PR and I'm not inclined to support it unless you can prove it's either the same or faster than the current method

**Comment by pigeonsk** [comment]
https://github.com/pigeonsk/miscdm/blob/master/test.dm

>#define INCREMENT 4

`32
31
31
32
34
30
32
32
31
33`
Average: 31.8

>#define INCREMENT (1<<2)

`32
34
29
33
31
31
34
30
32
32
`
Average:31.8

**Pull request merged**